### ISSUE TITLE
Find a decent allocation and initialization scheme for storage types

### ISSUE THREAD

**Issue by Robbepop** [paritytech/ink#16]
# The Problem

Storage types such as `storage::Value` can be used and created very differently.
It is required to initialized its wrapped inner value somehow, the exact procedure for that is depending on the usage scenario.

## Usage Scenarios

There are 3 common scenarios.

- **Deployment:** The contract is being deployed and has a `storage::Value`. Upon `deploy` the contract allocates the `storage::Value` on the correct location in the contract storage using `allocate_using` and the bump allocator. Afterwards it needs to initialize the allocated `storage::Value` to the value that the contract expects using `storage::Value`'s `OnDeploy` implementation. Most often users simply want the default value for this, e.g. `0` for `storage::Value<i32>`. Deployment requires allocation and initialization.

- **Call:** The contract is being called and has a `storage::Value`. As in deploy it needs to allocate it using the bump allocator. However, since it has already been initialized in the contract's `deploy` stage no further steps need to be taken and the `storage::Value` can be used safely.

- **Dynamic Creation:** Upon contract call or even upon deploy the contract might have the need to dynamically instantiate a `storage::Value` on the storage using the more complex `CellAndChunkAllocator`. For this the `storage::Value` not only needs to be allocated but also directly initialized using `OnDeploy`.

## Summary

| Scenario | Requires Allocation | Requires Initialization |
|:--|:--|:--|
| Deployment | Yes: Static | Yes |
| Call | Yes: Static | No |
| Dynamic | Yes: Dynamic | Yes |

## Solution

Ideally we have

```rust
struct Uninitialized<T>(T);
auto trait CanUseAfterAlloc {}
trait AllocateUsing {
    fn allocate_using(alloc: &mut Allocator) -> Uninitialized<Self>; // T: !CanUseAfterAlloc
    fn allocate_using(alloc: &mut Allocator) -> Self; // T: CanUseAfterAlloc
}
trait OnDeploy {
    fn on_deploy(&mut self, args: Self::Args)
}
```
Where `CanUseAfterAlloc` is not implemented if `Self` cannot be safely used after just allocation, such as `storage::Value`. Note that for example `storage::Vec` can be used after allocation depending on the exact implementation.
Also we need helper traits and routines to make life easier for the `Deployment` and `Dynamic` stages where we always need to allocate and initialize. Ideally this should be done in a single step.

**Comment by Robbepop** [comment]
Due to problems involving recursive calls to `allocate_using` we shifted the approach slightly.
The helper struct `MaybeUninitialized<T>` has been removed and `allocate_using` now returns `Self` instead of `MaybeUninitialized<Self>`.
Besides that we introduced another helper trait method `Initialize::initialize_into(self, Self::Args) -> Self` so that users can still use chaining.

## Summary

With these changes the following scenarios are approachable in the following ways.

- `CREATE`
    - `Foo::allocate_using(&mut bump_alloc).initialize_into(bar);` or
    - No code required to deploy the smart contract.
- `CALL`
    - `Foo::allocate_using(&mut bump_alloc);`
- `DYNAMIC`
    - `Foo::allocate_using(&mut cc_alloc).initialize(baz)` or
    - `Foo::allocate_using(&mut cc_alloc)` if no initialization is required

Note that all of this is going to be hidden from users normally as soon as the `pdsl_derive` is done.

**Comment by Robbepop** [comment]
To improve safety and security we should transit towards encapsulating the state and everything that modifies it.
For this we should separate state modifying environment functions out of `Env` into an own dedicated `State` trait.

This `State` trait could look like the following:

```rust
trait State {
    fn store(key: Key, bytes: &[u8]);
    fn load(key: Key) -> Option<Vec<u8>>;
    fn clear(key: Key);
}
```

It would be nice to have some compile-time niceties to check that a certain instance exists only once.
For runtime-based check there is https://crates.io/crates/once. With such a check we can secure that there is really just one state operating on contract state so that we can actually have single ownership of it. Maybe the future `eDSL` could provide this check to its users. Also note that the `State` trait is agnostic over `EnvTypes` and thus doesn't need to be generic over them.

The `Env` trait needs more polish. Types implementing it must also include in their state a type implementing the `State` trait. With this setup we can determine an ownership when using a single environment which is the default. Also the `Env` implementation could handle all dynamic and static allocations.

```rust
struct SrmlEnv<S: State> {
    state: S,
    bump_alloc: BumpAlloc,
    dyn_alloc: DynAlloc,
}
impl Env for SrmlEnv { ... }
```

Now we can actually form an API to allocate contract entities dynamically using an `Env` implementer for that purpose.

```rust
impl SrmlEnv {
    fn on_storage<T>(&mut self, args: <T as Initialize>::Args>) -> T
    where
        T: AllocateUsing + Initialize
    {
        T::allocate_using(&mut self.dyn_alloc).initialize_into(args)
    }
}
```

Note that dynamically allocated storage entities clear their associated storage upon `Drop` (going out of scope) whereas statically allocated storage entities do not. To fix the current situation for statically allocated entities we need to wrap them in a certain utility type in order to not clear their associated storage upon `Drop`.

**Comment by Robbepop** [comment]
Done (since ages). Closed.
